src/path.cr

```diff
@@ -0,0 +1,847 @@
+        end
+      end
+
+      if str.empty?
+        str << '.'
+      end
+
+      last_char = (str.buffer + str.bytesize - 1).value.unsafe_chr
+
+      if add_separator_at_end && !@type.separators.includes?(last_char)
+        str << @type.separator
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Yields each component of this path as a `String`.
+  def each_part
+    @type.each_part(@name) do |part|
+      yield part
+    end
+  end
+
+  # Returns the components of this path as an `Array(String)`.
+  def parts : Array(String)
+    parts = [] of String
+    each_part do |part|
+      parts << part
+    end
+    parts
+  end
+
+  # Converts this path to a Windows path.
+  def to_windows : Path
+    return self if windows?
+
+    new_instance(@name, Type::WINDOWS)
+  end
+
+  # Converts this path to a  POSIX path.
+  def to_posix : Path
+    return self if posix?
+
+    new_instance(@name.gsub(Type::WINDOWS.separator, Type::POSIX.separator), Type::POSIX)
+  end
+
+  # Converts this path to an absolute path. Relative paths are
+  # referenced from the current working directory of the process (`Path.current`)
+  # unless *base* is given, in which case it will be used as the reference path.
+  #
+  # ```
+  # Path["foo"].expand             # => Path["/current/path/foo"]
+  # Path["~/crystal/foo"].expand   # => Path["/home/crystal/foo"]
+  # Path["baz"].expand("/foo/bar") # => Path["/foo/bar/baz"]
+  # ```
+  def expand(base : Path | String = Path.current, home = Path.home) : Path
+    name = @name
+
+    if name.starts_with?('~')
+      home_ends_with_separator = @type.ends_with_separator?(home.@name)
+      home = new_instance(home.@name).normalize.to_s
+      # home = home + @type.separator if home_ends_with_separator && home.size > 1
+
+      if name.size == 1
+        return new_instance(home)
+      elsif @type.separators.includes?(name[1])
+        name = home + name[1..-1]
+      end
+    end
+
+    unless @type.starts_with_separator?(name)
+      base = new_instance(base.check_no_null_byte) unless base.is_a?(Path)
+      if name.empty?
+        name = base.expand.to_s
+      else
+        name = "#{base.expand}#{@type.separator}#{name}"
+      end
+    end
+
+    new_instance(name).normalize(remove_final_separator: false)
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")       # => Path["foo/bar/baz"]
+  # Path["foo/"].join("/bar/", "/baz")   # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/") # => Path["/foo/bar/baz/"]
+  # ```
+  def join(*parts) : Path
+    join parts
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")           # => Path["foo/bar/baz"]
+  # Path["foo/"].join(Path["/bar/", "/baz"]) # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/")     # => Path["/foo/bar/baz/"]
+  # ```
+  def join(parts : Array | Tuple) : Path
```

Comment:
Wouldn't `join(parts : Enumerable)` do?

Reply:
It looks like that's just been borrowed from `File.join`.